drive-stack/conUDS/src/main.rs

```diff
@@ -46,68 +50,220 @@ async fn main() {
+                        } else {
+                            error!("Unexpected DID length from node '{}': {:?}", node, node_crc_bytes);
+                        }
+                    }
+                    Err(e) => {
+                        error!("DID read (CRC) failed on node '{}': {:?}", node, e);
+                        // proceed with download anyway
+                    }
+                }
+
+                uds.client.start_persistent_tp().await;
+                let _ = uds.reset_node(SupportedResetTypes::Hard).await;
+
+                if let Err(e) = uds.file_download(&bin, 0x08002000).await {
+                    error!("Download failed for node '{}': {}", node, e);
+                }
+
+                uds.teardown().await;
+            }
+        }
+
+        // ===== single-node commands remain the same, but require -n =====
         ArgSubCommands::Reset(reset) => {
-            debug!(
-                "Performing {:#?} reset for node `{}`",
-                reset.reset_type, args.node
-            );
+            let node = args.node.clone().unwrap_or_else(|| {
+                error!("-n <node> is required for 'reset' (omit only when using 'batch').");
+                std::process::exit(1)
+            });
+            let uds_node = cfg.nodes.get(&node).unwrap_or_else(|| {
+                error!("UDS node '{}' not defined", node);
+                std::process::exit(1)
+            });
+
+            let mut uds = UdsSession::new(&args.device, uds_node.request_id, uds_node.response_id, true).await;
+            info!("Performing {:?} reset for node '{}'", reset.reset_type, node);
             if let Err(e) = uds.reset_node(reset.reset_type).await {
                 error!("Error while resetting ecu: {}", e);
             }
+            uds.teardown().await;
         }
+
         ArgSubCommands::Download(dl) => {
-            debug!(
-                "Downloading binary at '{:#?}' to node `{}`",
-                dl.binary, args.node
-            );
+            let node = args.node.clone().unwrap_or_else(|| {
+                error!("-n <node> is required for 'download' (omit only when using 'batch').");
+                std::process::exit(1)
+            });
+            let uds_node = cfg.nodes.get(&node).unwrap_or_else(|| {
+                error!("UDS node '{}' not defined", node);
+                std::process::exit(1)
+            });
+
+            let mut uds = UdsSession::new(&args.device, uds_node.request_id, uds_node.response_id, args.fast).await;
+            info!("Downloading binary {:?} to node '{}'", dl.binary, node);
+
+            if dl.verify_crc {
+                let mut file = File::open(dl.binary.to_path_buf()).expect("Binary does not exist!");
+                file.seek(SeekFrom::End(-4)).expect("Seek failed");
+                let mut buffer = [0u8; 4];
+                file.read_exact(&mut buffer).expect("CRC read failed");
+                let app_crc = u32::from_le_bytes(buffer);
```

Comment:
technically you should read the crc location from the app header at the start of the file instead of assuming it will be in the last four bytes. :footgun:

Reply:
Yea, this is a bit tricky for the bootloader verification case, but im technically not even supporting that presently. At this moment (and as I understand commonly) validation in added to the end of payloads so making this assumption here _seems_ fine